ha_mroonga.cpp

```diff
@@ -5671,7 +5671,13 @@ int ha_mroonga::wrapper_delete_table(const char *name,
     DBUG_RETURN(HA_ERR_OUT_OF_MEM);
   }
 
-#ifdef MRN_HANDLER_DELETE_TABLE_HAVE_TABLE_DEFINITION
+#ifdef MRN_HANDLER_DELETE_TABLE_HAVE_HTON_DROP_TABLE
+  if (ha_check_if_updates_are_ignored(ha_thd(), wrap_handlerton, "DROP")) {
```

Comment:
Why do we need this?

Reply:
https://github.com/MariaDB/server/commit/c55c292832e2776d37e06c43174ac006e00143a2#diff-a9025d3405668ea9267a0b3e4e109f7b doesn't do this.

Reply:
`ha_delete_table` function had called `ha_check_if_updates_are_ignored` so far.

See: https://github.com/MariaDB/server/commit/c55c292832e2776d37e06c43174ac006e00143a2#diff-d7b6eeffff29bb67048df7a850c034bdL4959

However, the function which is called in https://github.com/MariaDB/server/commit/c55c292832e2776d37e06c43174ac006e00143a2#diff-a9025d3405668ea9267a0b3e4e109f7bR5035 doesn't call `ha_check_if_updates_are_ignored`.

See: https://github.com/MariaDB/server/commit/c55c292832e2776d37e06c43174ac006e00143a2#diff-d7b6eeffff29bb67048df7a850c034bdR556-R566

`hton_drop_table` calles `my_delete` ( https://github.com/MariaDB/server/blob/10.5/mysys/my_delete.c#L31 ) finally.
I think that I can avoid the needless call of `unlink` by calling `ha_check_if_updates_are_ignored` in advance.

Reply:
Could you show call stack of `ha_mroonga::wrapper_delete_table()`?
If there is `handler::ha_drop_table()`, we don't need `ha_check_if_updates_are_ignored()` because it's already checked.

Reply:
[mysqld.1.trace.zst.tar.gz](https://github.com/mroonga/mroonga/files/5078848/mysqld.1.trace.zst.tar.gz)

I acquired a trace when I ran /mysql-test/mroonga/wrapper/drop_table/t/new_connection.test by using below command.

```
./test/run-sql-test.sh --debug --do-test=new_connection
```

`handler::ha_drop_table()` doesn't call in the avobe trace.

I acquired a trace file in the below environment.

* MariaDB 10.5.5
* Mroonga(master) + a below patch.

```diff
diff --git a/ha_mroonga.cpp b/ha_mroonga.cpp
index 3516a05a..c9e3c42c 100644
--- a/ha_mroonga.cpp
+++ b/ha_mroonga.cpp
@@ -5670,8 +5670,9 @@ int ha_mroonga::wrapper_delete_table(const char *name,
   {
     DBUG_RETURN(HA_ERR_OUT_OF_MEM);
   }
-
-#ifdef MRN_HANDLER_DELETE_TABLE_HAVE_TABLE_DEFINITION
+#ifdef MRN_HANDLER_DELETE_TABLE_HAVE_HTON_DROP_TABLE
+  error = wrap_handlerton->drop_table(wrap_handlerton, name);
+#elif defined(MRN_HANDLER_DELETE_TABLE_HAVE_TABLE_DEFINITION)
   error = hnd->ha_delete_table(name, table_def);
 #else
   error = hnd->ha_delete_table(name);
diff --git a/ha_mroonga.hpp b/ha_mroonga.hpp
index 3ea87cf5..659a477c 100644
--- a/ha_mroonga.hpp
+++ b/ha_mroonga.hpp
@@ -81,6 +81,10 @@ extern "C" {
 #  define MRN_HANDLER_HAVE_SET_HA_SHARE_REF
 #endif
 
+#if (MYSQL_VERSION_ID >= 100505) && defined(MRN_MARIADB_P)
+#  define MRN_HANDLER_DELETE_TABLE_HAVE_HTON_DROP_TABLE
+#endif
+
 #if MYSQL_VERSION_ID >= 50706 && !defined(MRN_MARIADB_P)
 #  define MRN_BIG_TABLES
 #elif defined(BIG_TABLES)
```

Reply:
Thanks.
I also confirmed with `bt` in GDB. `handler::ha_drop_table()` isn't really called:

```text
(gdb) bt
#0  0x00007ffff30e95d0 in ha_mroonga::wrapper_delete_table(char const*, handlerton*, char const*)@plt ()
   from /home/kou/work/cpp/mariadb-10.5.5.build/mysql-test/var/plugins/ha_mroonga.so
#1  0x00007ffff30f9bf2 in ha_mroonga::delete_table (this=0x7fffa40133d0, name=0x7fffd8b61820 "./test/logs") at ../mroonga/ha_mroonga.cpp:5806
#2  0x00005555561ac1b2 in hton_drop_table (hton=0x555558239eb8, path=0x7fffd8b61820 "./test/logs") at /home/kou/work/cpp/mariadb-10.5.5/sql/handler.cc:564
#3  0x00005555561b18a8 in ha_delete_table (thd=0x7fffa4000e78, hton=0x555558239eb8, path=0x7fffd8b61820 "./test/logs", db=0x7fffd8b616a0, 
    alias=0x7fffd8b616b0, generate_warning=true) at /home/kou/work/cpp/mariadb-10.5.5/sql/handler.cc:2765
#4  0x0000555555f41df6 in mysql_rm_table_no_locks (thd=0x7fffa4000e78, tables=0x7fffa4012ab8, if_exists=false, drop_temporary=false, drop_view=false, 
    drop_sequence=false, dont_log_query=false, dont_free_locks=false) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_table.cc:2482
#5  0x0000555555f40f1c in mysql_rm_table (thd=0x7fffa4000e78, tables=0x7fffa4012ab8, if_exists=false, drop_temporary=false, drop_sequence=false, 
    dont_log_query=false) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_table.cc:2143
#6  0x0000555555e5acab in mysql_execute_command (thd=0x7fffa4000e78) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_parse.cc:4922
#7  0x0000555555e64ccb in mysql_parse (thd=0x7fffa4000e78, rawbuf=0x7fffa4012a00 "DROP TABLE logs", length=15, parser_state=0x7fffd8b62510, 
    is_com_multi=false, is_next_command=false) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_parse.cc:7994
#8  0x0000555555e511a9 in dispatch_command (command=COM_QUERY, thd=0x7fffa4000e78, packet=0x7fffa4008fd9 "DROP TABLE logs", packet_length=15, 
    is_com_multi=false, is_next_command=false) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_parse.cc:1867
#9  0x0000555555e4f9ce in do_command (thd=0x7fffa4000e78) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_parse.cc:1348
#10 0x0000555555ff4162 in do_handle_one_connection (connect=0x55555969d258, put_in_cache=true)
    at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_connect.cc:1410
#11 0x0000555555ff3ed4 in handle_one_connection (arg=0x55555969d258) at /home/kou/work/cpp/mariadb-10.5.5/sql/sql_connect.cc:1312
#12 0x00005555564f4a20 in pfs_spawn_thread (arg=0x55555969d338) at /home/kou/work/cpp/mariadb-10.5.5/storage/perfschema/pfs.cc:2201
#13 0x00007ffff7aa2ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
#14 0x00007ffff769fdaf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

I checked MariaDB's code. `ha_check_if_updates_are_ignored()` is introduced since MariaDB 10.5. And it's for S3 storage engine. `s3_slave_ignore_updates` variable https://mariadb.com/kb/en/s3-storage-engine-system-variables/#s3_slave_ignore_updates is related.

It seems that `ha_drop_table()` (not `handler::ha_drop_table()`) in https://github.com/MariaDB/server/commit/c55c292832e2776d37e06c43174ac006e00143a2 isn't used anywhere. I'm not sure it's a bug of MariaDB or not.

I think that we should not touch `ha_check_if_updates_are_ignored()` from Mroonga like MariaDB does. It should be processed in MariaDB.

How about use the patch in https://github.com/mroonga/mroonga/pull/334/files#r470996804 instead of the pull request for now? If we receive a bug report, we can revisit this.

Reply:
> I think that we should not touch `ha_check_if_updates_are_ignored()` from Mroonga like MariaDB does. It should be processed in MariaDB.
> 
> How about use the patch in https://github.com/mroonga/mroonga/pull/334/files#r470996804 instead of the pull request for now? If we receive a bug report, we can revisit this.

I see.
I agree with you. I use the patch in https://github.com/mroonga/mroonga/pull/334/files#r470996804 instead of the pull request.